runtime/memprof.c

```diff
@@ -518,10 +665,85 @@ void caml_memprof_track_interned(header_t* block, header_t* blockend) {
       p = next_p;
     }
 
-    register_postponed_callback(
-      Val_hp(p), mt_generate_binom(next_p - next_sample_p) + 1,
-      Unmarshalled, &callstack);
-
+    if(callstack == 0) callstack = capture_callstack_postponed();
+    if(callstack == 0) break;  /* OOM */
+    new_tracked(mt_generate_binom(next_p - next_sample_p) + 1,
+                Hd_hp(p), 1, young, Val_hp(p), callstack);
     p = next_p;
   }
+  caml_memprof_check_action_pending();
+}
+
+/**** Interface with the OCaml code. ****/
+
+CAMLprim value caml_memprof_set(value lv, value szv,
+                                value cb_alloc_minor, value cb_alloc_major,
+                                value cb_promote,
+                                value cb_dealloc_minor, value cb_dealloc_major)
+{
+  CAMLparam5(lv, szv, cb_alloc_minor, cb_alloc_major, cb_promote);
+  CAMLxparam2(cb_dealloc_minor, cb_dealloc_major);
+  double l = Double_val(lv);
+  intnat sz = Long_val(szv);
+  uintnat i;
+
+  if (sz < 0 || !(l >= 0.) || l > 1.) /* Checks that [l] is not NAN. */
+    caml_invalid_argument("caml_memprof_set");
+
+  /* This call to [caml_memprof_set] will discard all the previously
+     tracked blocks. We try one last time to call the postponed
+     callbacks. */
+  caml_raise_if_exception(caml_memprof_handle_postponed_exn());
+
+  /* Discard the tracked blocks. */
+  for(i = 0; i < tracked.len; i++)
+    if(tracked.entries[i].idx_ptr != NULL)
+      *tracked.entries[i].idx_ptr = Invalid_index;
+  tracked.len = 0;
+  tracked.callback = tracked.young = tracked.delete = 0;
+  caml_stat_free(tracked.entries);
+  tracked.entries = NULL;
+  tracked.alloc_len = 0;
+
+  if (!init) {
```

Comment:
The OCaml runtime now supports a `caml_shutdown` function which should allow one to shutdown and then restart a runtime instance from the same process. The present logic may not handle this properly. We think it would be cleaner to have separate `caml_memprof_init` and `caml_memprot_shutdown` functions, and ensure that `caml_memprof_shutdown` is called from `caml_shutdown` (and sets this `init` flag back to `0`).